module/zfs/dbuf.c

```diff
@@ -1087,9 +1100,13 @@ dbuf_read_impl(dmu_buf_impl_t *db, zio_t *zio, uint32_t flags)
 
 	dbuf_add_ref(db, NULL);
 
+	zio_flags = (flags & DB_RF_CANFAIL) ?
+	    ZIO_FLAG_CANFAIL : ZIO_FLAG_MUSTSUCCEED;
+	if ((flags & DB_RF_NO_DECRYPT) && BP_IS_ENCRYPTED(db->db_blkptr))
+		zio_flags |= ZIO_FLAG_RAW;
```

Comment:
> Is this used for blocks of dnodes, other blocks, or both?

Both.

> If used on a block of dnodes, do we really want RAW == RAW_COMPRESS | RAW_ENCRYPT? I.e. do we want to get the compressed data?

Since encryption is applied after compression it is impossible to decompress anything without decrypting it first. I structured the code in such a way that `ZIO_FLAG_RAW_ENCRYPT` must always imply `ZIO_FLAG_RAW_COMPRESS`, even if compression is off so that callers don't expect encrypted but decompressed data.